### PR TITLE
swarm: make smart-dialing opt in

### PR DESCRIPTION
**Pull request by sukunrt**

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -124,7 +124,8 @@ type Config struct {
 	DisableMetrics       bool
 	PrometheusRegisterer prometheus.Registerer
 
-	NoDelayNetworkDialRanker bool
+	SmartDialing       bool
```
Comment (marten-seemann) [inline config/config.go:127]: Why is this a bool? How would a user set a different dial ranker than the default one.

File: defaults.go [hunk 2]
```diff
@@ -135,6 +135,11 @@ var DefaultPrometheusRegisterer = func(cfg *Config) error {
 	return cfg.Apply(PrometheusRegisterer(prometheus.DefaultRegisterer))
 }
 
+// DefaultSmartDialing configures libp2p to disable smart dialing by default
+var DefaultSmartDialing = func(cfg *Config) error {
```
[the remark this answers is not in the document]
Comment (sukunrt) [reply defaults.go]: I don't get this. This helps us flip the default option in a later release. Is there a way to do this without this?

File: config/config.go (hunk 1 above)
Comment (sukunrt) [reply config/config.go]: I made it a yes no options. Have changed this to allow custom dial ranker implementations.

**Comment by sukunrt** [comment]
@marten-seemann ignoring the changelog, is this what you had in mind?

File: defaults.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (marten-seemann) [reply defaults.go]: Do we need to explicitly define a type? Can’t we just apply the default if no dial ranker is set?

[the remark this answers is not in the document]
Comment (sukunrt) [reply defaults.go]: To be clear we are now discussion this default config fields: `DefaultDialRanker` and `DefaultDialRankerCustom`. 

If we don't have `DialRankerCustom` we cannot distinguish between cases where the user has not provided a value and where the user has explicitly provided nil to disable smart dialing. This is not an issue now but for the future release where we want to turn smart dialing on by default we will need to use `swarm.DefaultDialRanker` in such cases. I want to keep the behaviour that option `DialRanker(nil)` disables smart dialing in future releases. 

Using a Default Option just felt cleaner. see `DefaultEnableRelay`

[the remark this answers is not in the document]
Comment (marten-seemann) [reply defaults.go]: What about doing the thing we did for the rcmgr? Provide a `NullDialRanker` implementation. iiuc we already have that, the `NoDelayDialRanker`.

[the remark this answers is not in the document]
Comment (sukunrt) [reply defaults.go]: I don't have strong opinions on this. Removed the field `DialRankerCustom` and the explicit default option. Now default is set here https://github.com/sukunrt/go-libp2p/blob/smart-dialing-optin/config/config.go#L180

**Comment by MarcoPolo** [comment]
No strong opinion, but I think it's a nicer api for the user to be able to enable/disable smart dialing with a boolean rather than having users know about dial rankers.

**Review by MarcoPolo — APPROVED** [review]

**Pull request merged**